### ISSUE TITLE
Normalization issue in Molden writer when using Cartesian basis functions

### ISSUE THREAD

**Issue by tovrstra** [psi4/psi4#1973]
Example: [molden_cartesian.zip](https://github.com/psi4/psi4/files/4988043/molden_cartesian.zip)

The ZIP file contains a trivial HF/6-31G* calculation on water with PSI4 and Gaussian. To allow for a one-on-one comparison, symmetry, reorientation etc. are disabled. The molden file written out by PSI4 results in the wrong total charge when it is loaded into Molden, which shows there is a problem:

```
                        Mulliken Charges
                        ================
 
               O  -0.8774
               H   0.4308
               H   0.4316
 
Sum of Mulliken Charges:  -0.0150
```

Repeating the calculation in Gaussian and loading the log file into Molden gives:

```
                        Mulliken Charges
                        ================
 
               O  -0.8652
               H   0.4323
               H   0.4329
 
Sum of Mulliken Charges:  -0.0000
```

I've written out the orbitals from the Gaussian log back into a Molden file with Molden, to allow for a one-on-one comparison of the orbital coefficients. This revealed that coefficients 13, 14 and 15 of each orbital in the Molden file written by PSI4 are a factor **sqrt(3)** too large compared to what they should be. These are coefficients for the `xy`, `xz` and `yz` basis functions in a d shell. Other d functions are fine. I'll check for g and f functions as well...

**Comment by tovrstra** [comment]
Here is another example: [molden_cartesian2.zip](https://github.com/psi4/psi4/files/4988357/molden_cartesian2.zip)

Comparing orbital coefficients, it seems that several of them are too large compared to what they should be for the Molden format:

- The 3 last coefficients of the Cartesian d shell: `xy`, `xz`, `yz`: factor **sqrt(3)** too large.
- Coefficients 4, 5, 6, 7, 8 and 9 of the Cartesian f shell: `xyy`, `xxy`, `xxz`, `xzz`, `yzz`, `yyz`: factor **sqrt(5)** too large.
- Coefficient 10 of the Cartesian f shell (last one): `xyz`: factor **sqrt(15)** too large.
- Coefficients 4, 5, 6, 7, 8,  and 9 of the Cartesian g shell: `xxxy`, `xxxz`, `yyyx`, `yyyz`, `zzzx`, `zzzy`: factor **sqrt(7)** too large.
- Coefficients 10, 11, 12 of the Cartesian g shell: `xxyy`, `xxzz`, `yyzz``: factor **sqrt(35/3)** too large.
- Coefficients 13, 14, 15 of the Cartesian g shell: `xxyz`, `yyxz`, `zzxy`: factor **sqrt(35)** too large.

(My initial post contains a few mistakes, which I'll fix right away.)

**Comment by andysim** [comment]
Thanks for the report.  When I put together the first version of the FCHK writer, [I accounted for](https://github.com/psi4/psi4/blob/master/psi4/src/psi4/libmints/writer.cc#L395-L472) the fact that Psi4 uses the CCA standard, but these normalizations didn't get applied to the Molden writer it seems.  This would be a very good issue to tackle for anybody looking for a good beginner project; the comments in the FCHK writer should serve as a useful guide and I'd be very happy to consult.  If there are no volunteers in a week, I'll apply the changes.

**Comment by JonathonMisiewicz** [comment]
If anybody feels like moving the molden writer Py-side while they're at it, the [NBO writer](https://github.com/JonathonMisiewicz/psi4/blob/master/psi4/driver/p4util/writer.py) should be another good reference. No deep understanding of normalization conventions required.

**Comment by andysim** [comment]
Excellent point, @JonathonMisiewicz - this is much better suited to the Python layer.

**Comment by tovrstra** [comment]
I will leave the fix in PSI4 to someone else if you don't mind. I've added a workaround for this issue in our Molden parser. Feel free to take a look: https://github.com/theochem/iodata/blob/596cac8db3d3795eaed704b35c99b71b25447bc1/iodata/formats/molden.py#L506-L536 

If someone wants to code a Python version of the Molden writer, please do so with regression tests. I've gone through some trouble to fix small bugs in the C++ code in previous PRs and it would be a shame to reintroduce similar bugs in a Python rewrite. (It is easy to introduce small mistakes.) In general, to test the generated molden files, one can always load them into the Molden program and check if the total charge is correct, as well as the Mulliken charges, which can be compared to the output of `psi4.oeprop(wfn, "MULLIKEN_CHARGES")`. This can be verified with increasing angular momentum of the AO basis and pure versus Cartesian conventions.

**Comment by sgoodlett** [comment]
I will be working on this issue with @JonathonMisiewicz

**Comment by tovrstra** [comment]
@sgoodlett That's great. If you need me to test things, please let me know.

**Comment by JonathonMisiewicz** [comment]
@sgoodlett and I have been working on this and have to ask about an older PR. @tovrstra, in #1510, why did you go to **17** decimal places? I understand that we can go beyond four decimal places, but I wouldn't expect reproducibility to 17 decimal places due to finite machine precision, and indeed, we're seeing errors out that far.

**Comment by tovrstra** [comment]
The large number of digits was introduced to eliminate round-off errors completely. There are a few reasons:

- When testing software that post-processes the wavefunction, it is desirable to be able to validate the numerical algorithms in such codes. For example, this can be done by checking the total number of electrons on very fine integration grids in such post-processing codes. When doing so, all other sources of errors should be eliminated to avoid confusion. We noticed that this was especially problematic before #1510 with fractionally occupied natural orbitals.
- When deriving approximate electrostatic models from a wavefunction, satisfying charge neutrality is quite important for long-range interactions. I know you can always "fix" this *a posteriori* (e.g. by shifting partial charges to match the net charge), but avoiding the need for a fix is still preferable, just to rule out sources of error.

So why 17? The number of significant digits from a double precision number varies between 15 and 17. See https://en.wikipedia.org/wiki/Double-precision_floating-point_format (For some numbers, digits 16 and/or 17 may be insignificant, but not always.)

**Comment by JonathonMisiewicz** [comment]
@sgoodlett, it's been nearly a year. Do you want to finish this, or are you leaving this for me to take over?

**Comment by JonathonMisiewicz** [comment]
Closed by #2229.
